Xcode 4 support

jverkoey opened the pull request:
This gives us full support for Xcode 4. The best part is that devs with Xcode 3.2.# who wish to upgrade to Xcode 4 just have to run the ttmodule script again!

For those who don't want to/can't use ttmodule, you simply need to add the following to your project's `HEADER_SEARCH_PATHS`:

```
"$(BUILT_PRODUCTS_DIR)/../three20"
"$(BUILT_PRODUCTS_DIR)/../../three20"
```

I'll put together a quick migration article on Three20.info outlining this for historical purposes.

Commit:
[scripts] Xcode 4 support to ttmodule.py

jverkoey committed:
[scripts] Check for the existence of an extra folder in the Protect command.

Xcode 4 places an extra folder named "ArchiveIntermediates" in the path
when doing an Archive build. Before this change, this would cause the
Protect script to fail when doing an Archive build because we weren't
stepping up far enough in the directory tree.

jverkoey committed:
[config] Turn on the SKIP_INSTALL option for libraries to enable support for IPA building with Xcode 4.

jverkoey committed:
[config] Add Xcode 4 header search paths that take into account the new Build Location value.

Before this change, Three20 compilation would fail whenever private
headers were imported because the headers were being copied to a new
location.

jverkoey committed:
[scripts] Trim quotes from existing linker flag value so we don't duplicate linker flags on subsequent ttmodule runs.

jverkoey committed:
[scripts] Clean up the help output for ttmodule.py.

jverkoey committed:
[scripts] Add version support to Pbxproj and use it to determine whether we need to add special settings for Xcode 4.

pazustep commented on src/scripts/Protect.command:
```diff
@@ -20,7 +20,13 @@
-cd ${CONFIGURATION_BUILD_DIR}${PUBLIC_HEADERS_FOLDER_PATH}
+# In Xcode 4 Archive builds, there is one extra folder placed between the configuration build
+# dir and the public headers path titled "ArchiveIntermediates".
+if [[ $CONFIGURATION_BUILD_DIR == *ArchiveIntermediates* ]]; then
```
Maybe you should test for ${DEPLOYMENT_LOCATION} == "YES" to detect an archive run instead of testing the folder name — Apple, or maybe the user, can change the built path to something that doesn't contain "ArchivedIntermediates".

jverkoey replied:
I had originally used that but then wasn't sure if it would do the same on Xcode 3.2.5. Do you know if that is an Xcode 4-specific property?

pazustep replied:
I've seen references to DEPLOYMENT_LOCATION on Google back on 2005, so I'm pretty sure this wasn't introduced in Xcode 4. On the other hand, I didn't test it on Xcode 3.2, so it might behave differently there. In this case, testing XCODE_VERSION_MAJOR as well might be a better idea.

jverkoey replied:
Hm. I guess if we do a combination of `XCODE_VERSION_MAJOR` and `DEPLOYMENT_LOCATION` we'll be pretty safe.

jverkoey replied:
Fixed in a new commit.

https://github.com/facebook/three20/pull/466

jverkoey committed:
[scripts] Add --xcode-version to ttmodule in order to override the project version.

This is necessary for those wanting to upgrade their Three20 settings
in a Xcode 3.2.5 project to Xcode 4. Before, the script would look
at the project version and see that it was for 3.2.5 and only add the
3.2.5 settings. Now, it's possible to state --xcode-version=4 and this
will force ttmodule to treat the project file as though it were version
4.

jverkoey committed:
[scripts] Use a less hacky check to determine if we're doing an Xcode 4 Archive build.

jwang commented on src/scripts/ttmodule.py:
```diff
@@ -107,40 +108,52 @@ def main():
+    Print all dependencies for the Three20UI module
+    > %prog -d Three20UI
+    
+    Print all dependencies for the Three20 module's Three20-Xcode3.2.5 target.
```
these don't exist anymore. That's version specific junk from previous releases.

jverkoey replied:
Hm, fair point. I'll remove it.

jverkoey replied:
Done.

jverkoey committed:
[scripts] Remove version specific help information from ttmodule.

jverkoey commented:
Closed by a8787dc95d2e50ac111f49c75a381b6852100b27.
